moto/cloudwatch/models.py

```diff
@@ -178,16 +183,19 @@ def __init__(self, namespace, name, value, dimensions, timestamp, unit=None):
         ]
         self.unit = unit
 
-    def filter(self, namespace, name, dimensions, already_present_metrics):
+    def filter(self, namespace, name, dimensions, already_present_metrics=[]):
         if namespace and namespace != self.namespace:
             return False
         if name and name != self.name:
             return False
-        for metric in already_present_metrics:
-            if self.dimensions and are_dimensions_same(
-                metric.dimensions, self.dimensions
-            ):
-                return False
+
+        # Note: The logic below appears to be incorrect. We should not skip adding new Metrics
```

Comment:
What is the reason behind the removal of this logic?
It was added to fix this issue: https://github.com/localstack/localstack/issues/3222

Note that in AWS itself, this scenario would never occur - as the put_metric_data() call fails when trying to store a duplicate metric.
However, as we (moto) do not have this validation yet, we do have to deal with potentially dirty data. So it makes sense IMO to at least keep the outgoing data consistent with what AWS would return

Reply:
Good catch, thanks @bblommers . This was part of a general overhaul of the CW metrics logic - we've been working with @usmangani1 (he's part of our team) on [this PR](https://github.com/spulec/moto/pull/3461) that you've kindly helped review, and later discovered that a few more changes were needed.

Just double-checked against the latest version of LocalStack (which contains the code in this PR), and the behavior reported by the user in localstack/localstack#3222 still seems to be accurate:

```
$ awslocal cloudwatch list-metrics --namespace AWS/EC2 --dimensions Name=InstanceId,Value=i-0bada24f4d28a7793
{
    "Metrics": []
}
$ awslocal cloudwatch put-metric-data --namespace AWS/EC2 --metric-data '[{ "MetricName": "CPUUtilization", "Dimensions": [ { "Name": "InstanceId", "Value": "i-0bada24f4d28a7793" } ], "Value": 28, "Unit": "Percent" }]'
$ awslocal cloudwatch list-metrics --namespace AWS/EC2 --dimensions Name=InstanceId,Value=i-0bada24f4d28a7793
{
    "Metrics": [
        {
            "Namespace": "AWS/EC2",
            "MetricName": "CPUUtilization",
            "Dimensions": [
                {
                    "Name": "InstanceId",
                    "Value": "i-0bada24f4d28a7793"
                }
            ]
        }
    ]
}
$ awslocal cloudwatch put-metric-data --namespace AWS/EC2 --metric-data '[{ "MetricName": "CPUUtilization", "Dimensions": [ { "Name": "InstanceId", "Value": "i-0bada24f4d28a7793" } ], "Value": 28, "Unit": "Percent" }]'
$ awslocal cloudwatch list-metrics --namespace AWS/EC2 --dimensions Name=InstanceId,Value=i-0bada24f4d28a7793
{
    "Metrics": [
        {...}  # same as above ...
    ]
}
```

Hope that makes sense. Happy to discuss more and/or work with you on adding more tests to cover different edge cases for CloudWatch Metrics (as well as adding any missing validation) post merge. 👍

Reply:
Hmm. I wonder if the AWS behavior changed then recently, because the error that duplicate dimensions are invalid definitely didn't show up when I was testing this exact scenario in #3461

Reply:
Good catch - commented the code back in and re-enabled the disabled test (also added a new test to cover get_metric_statistics using metrics with different dimensions).